pkg/controller/node/cidr_allocator.go

```
@@ -0,0 +1,182 @@
+limitations under the License.
+*/
+
+package node
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+)
+
+const (
+	// cidrsBufSize is the buffer size of the free CIDRs channel
+	cidrsBufSize = 16
+	// nextFreeCIDRCooldown is the cooldown time before a released CIDR could be reused
+	nextFreeCIDRCooldown = 500 * time.Millisecond
+	// allocateNextTimeout control how long AllocateNext() will block before it return an error
+	allocateNextTimeout = 1 * time.Second
+)
+
+var errCIDRRangeNoCIDRsRemaining = errors.New("CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
+
+// CIDRAllocator is an interface implemented by things that know how to allocate/recycle CIDR for nodes.
+type CIDRAllocator interface {
+	AllocateNext() (*net.IPNet, error)
+	Release(*net.IPNet)
+}
+
+// emptyAllocator is a CIDRAllocator implementation that allocates empty CIDRs, yielding
+// a CIDR value of nil for every call to AllocateNext().
+type emptyAllocator struct{}
+
+// AllocateNext always returns nil and error
+func (r *emptyAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	return nil, errCIDRRangeNoCIDRsRemaining
+}
+
+// Release is a noop
+func (r *emptyAllocator) Release(_ *net.IPNet) {
+	// noop
+}
+
+type rangeAllocator struct {
+	clusterCIDR    *net.IPNet
+	subNetMaskSize int
+	cidrs          chan *net.IPNet
+	used           big.Int
+	lock           sync.Mutex
+	rand           *rand.Rand
+}
+
+// NewCIDRRangeAllocator returns a CIDRAllocator to allocate CIDR for node
+// If subNetMaskSize is less than cluster CIDR mask size, an emptyAllocator is returned.
+// Otherwise an rangeAllocator is returned.
+func NewCIDRRangeAllocator(clusterCIDR *net.IPNet, subNetMaskSize int) CIDRAllocator {
+	cidrMask := clusterCIDR.Mask
+	cidrMaskSize, _ := cidrMask.Size()
+
+	if subNetMaskSize < cidrMaskSize {
+		return &emptyAllocator{}
```

Did you have a reason that you didn't want to return an error from this function?